types/wlr_input_method.c

```diff
@@ -0,0 +1,212 @@
+﻿#ifndef _POSIX_C_SOURCE
+#define _POSIX_C_SOURCE 200809L
+#endif
+#include <assert.h>
+#include <string.h>
+#include <wayland-util.h>
+#include <wlr/types/wlr_input_method.h>
+#include <wlr/types/wlr_surface.h>
+#include <wlr/util/log.h>
+#include <xkbcommon/xkbcommon.h>
+#include "input-method-unstable-v2-protocol.h"
+#include "util/signal.h"
+
+static void noop() {}
+
+static void commit(struct wl_client *client, struct wl_resource *resource,
+		uint32_t serial) {
+	struct wlr_input_method *context =
+		wl_resource_get_user_data(resource);
+	context->current = context->pending;
+	context->current_serial = serial;
+	struct wlr_input_method_state default_state = {0};
+	context->pending = default_state;
+	wlr_signal_emit_safe(&context->events.commit, (void*)context);
+}
+
+static void commit_string(struct wl_client *client,
+		struct wl_resource *resource, const char *text) {
+	struct wlr_input_method *context =
+		wl_resource_get_user_data(resource);
+	free(context->pending.commit_text);
+	context->pending.commit_text = strdup(text); // TODO: clean up
+}
+
+static void preedit_string(struct wl_client *client,
+		struct wl_resource *resource, const char *text, int32_t cursor_begin,
+		int32_t cursor_end) {
+	struct wlr_input_method *context =
+		wl_resource_get_user_data(resource);
+	context->pending.preedit.cursor_begin = cursor_begin;
+	context->pending.preedit.cursor_end = cursor_end;
+	free(context->pending.preedit.text);
+	context->pending.preedit.text = strdup(text); // TODO: clean up
+}
+
+static void delete_surrounding_text(struct wl_client *client,
+		struct wl_resource *resource, uint32_t before_length,
+		uint32_t after_length) {
+	struct wlr_input_method *context =
+		wl_resource_get_user_data(resource);
+	context->pending.delete.before_length = before_length;
+	context->pending.delete.after_length = after_length;
+}
+
+static const struct zwp_input_method_v2_interface input_method_impl;
+
+static struct wlr_input_method *input_method_from_resource(
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource,
+		&zwp_input_method_v2_interface, &input_method_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+static const struct zwp_input_method_v2_interface input_method_impl = {
+	.destroy = noop,
+	.commit = commit,
+	.commit_string = commit_string,
+	.preedit_string = preedit_string,
+	.delete_surrounding_text = delete_surrounding_text,
+	.get_input_popup_surface = noop,
+	.grab_keyboard = noop,
+};
+
+static void input_method_destroy(struct wlr_input_method *context) {
+	wlr_signal_emit_safe(&context->events.destroy, context);
+	// the resource must be destroyed after this call by the client, so proceed with freeing everything else
+	wl_resource_set_user_data(context->resource, NULL);
+	free(context);
+}
+
+static void input_method_handle_resource_destroy(struct wl_resource *resource) {
+	struct wlr_input_method *context = input_method_from_resource(resource);
+	if (!context) {
```

Comment:
`input_method_destroy` should check for NULL

Reply:
No, but that's how inert resources work. See the last part of `CONTRIBUTING.md`.

Reply:
Thanks, I didn't realize this guide was there.